### PR TITLE
RSDK-6185 Send/receive module logs over gRPC

### PR DESCRIPTION
**Pull request by benjirewis**
Adds client/server wrappers for `Log`. Uses those wrappers respectively to log from modules through a custom `Appender` and to receive module logs and output them through RDK logs.

### CODE REVIEW

File: examples/customresources/demos/simplemodule/module.go
Comment (benjirewis) [inline examples/customresources/demos/simplemodule/module.go:60]: Logs above would look like:
```
2023-12-11T18:03:04.797-0500    INFO    robot_server    server/server.go:413    simple-module: info message
2023-12-11T18:03:04.797-0500    WARN    robot_server    server/server.go:415    simple-module: warn message
2023-12-11T18:03:04.797-0500    ERROR   robot_server    server/server.go:417    simple-module: error message
```
When received by a (non-debug) server.

And like:
```
2023-12-11T18:15:13.928-0500	INFO	robot_server.process.SimpleModule_/Users/benji.rewis/Code/rdk/examples/customresources/demos/simplemodule/simplemodule.StdOut	pexec/managed_process.go:244	
\_ 2023-12-11T23:15:13.928Z	INFO	simple-module	simplemodule/module.go:57	info message
2023-12-11T18:15:13.928-0500	INFO    robot_server.process.SimpleModule_/Users/benji.rewis/Code/rdk/examples/customresources/demos/simplemodule/simplemodule.StdOut	pexec/managed_process.go:244	
\_ 2023-12-11T23:15:13.928Z	WARN	simple-module	simplemodule/module.go:58	warn message
2023-12-11T18:15:13.928-0500	INFO	robot_server.process.SimpleModule_/Users/benji.rewis/Code/rdk/examples/customresources/demos/simplemodule/simplemodule.StdOut	pexec/managed_process.go:244	
\_ 2023-12-11T23:15:13.928Z	ERROR	simple-module	simplemodule/module.go:59	error message
```

When sent to the module's STDOUT due to an inactive or unhealthy gRPC connection.

I don't love the `server/server.go:41*` pieces of the RDK logs.

File: go.mod
Comment (benjirewis) [inline go.mod:401]: To use https://github.com/viamrobotics/api/pull/414.

File: robot/client/client.go
Comment (benjirewis) [inline robot/client/client.go:941]: It's possible to not make this public API if we're really concerned with Golang SDK users (or module authors) improperly using it. It was easier for me to expose a new method here, though.

**Comment by benjirewis** [comment]
cc @maximpertsov and @cheukt (would be great to have @viamrobotics/netcode 🤔) for NetCode interest.

File: examples/customresources/demos/simplemodule/module.go
Comment (benjirewis) [inline examples/customresources/demos/simplemodule/module.go:60]: These logs now look like:
```
2023-12-13T18:01:54.941-0500    ERROR   robot_server.simple-module      simplemodule/module.go:59 "error message" {"module_log_ts": Dec 13 23:01:54, myKey: myVal}
2023-12-13T18:01:55.943-0500    INFO    robot_server.simple-module      simplemodule/module.go:57 "info message" {"module_log_ts": Dec 13 23:01:55, myKey: myVal}
2023-12-13T18:01:55.944-0500    WARN    robot_server.simple-module      simplemodule/module.go:58 "warn message" {"module_log_ts": Dec 13 23:01:55, myKey: myVal}
```
When received by a non-debug server.

And like:

```
2023-12-11T18:15:13.928-0500	INFO	robot_server.process.SimpleModule_/Users/benji.rewis/Code/rdk/examples/customresources/demos/simplemodule/simplemodule.StdOut	pexec/managed_process.go:244	
\_ 2023-12-11T23:15:13.928Z	INFO	simple-module	simplemodule/module.go:57	info message
2023-12-11T18:15:13.928-0500	INFO    robot_server.process.SimpleModule_/Users/benji.rewis/Code/rdk/examples/customresources/demos/simplemodule/simplemodule.StdOut	pexec/managed_process.go:244	
\_ 2023-12-11T23:15:13.928Z	WARN	simple-module	simplemodule/module.go:58	warn message
2023-12-11T18:15:13.928-0500	INFO	robot_server.process.SimpleModule_/Users/benji.rewis/Code/rdk/examples/customresources/demos/simple-module/simplemodule.StdOut	pexec/managed_process.go:244	
\_ 2023-12-11T23:15:13.928Z	ERROR	simple-module	simplemodule/module.go:59	error message
```
When sent to the module's STDOUT due to an inactive or unhealthy gRPC connection.

File: robot/client/client.go
Comment (benjirewis) [inline robot/client/client.go:957]: This field formatting and insertion of the module log timestamp is hacky... I couldn't find a great way around this, and I found it easier to stringify things on the client side rather than handle everything on the server side after converting to and from proto.

File: robot/server/server.go
Comment (benjirewis) [inline robot/server/server.go]: I started creating `Sublogger`s in the `ModuleLog` receiver on the server-side instead of a global one here. I also did _not_ set a DEBUG log level for them, was this line intentional or just for testing, @dgottlieb?

File: logging/impl.go
Comment (benjirewis) [inline logging/impl.go:69]: `impl` `Sublogger`s were not retaining UTC value.

File: robot/server/server.go
Comment (benjirewis) [inline robot/server/server.go:418]: @dgottlieb even with my change of passing `inUTC` through to subloggers, the `logger` created here will still be in EST (local time) because `WithOptions` converts to an actual zap logger and we lose the UTCness of the sublogger 😢. Not sure what to do about that, but reining in all our log timezones is probably out of scope for this change anyway?

File: examples/customresources/demos/simplemodule/module.go
Comment (dgottlieb) [reply examples/customresources/demos/simplemodule/module.go:60]: Looks great! I left a comment in the string formatting below. And thanks for the examples so I didn't have to guess if the string format was as I expected.

**Review by dgottlieb — CHANGES_REQUESTED** [review]

File: robot/server/server.go
Comment (dgottlieb) [reply robot/server/server.go]: This line was intentional. Otherwise a module set to debug level would send an module log GRPC request, but the `r.Logger` would filter it out.

Comment (dgottlieb) [reply robot/server/server.go]: (I should have documented the intention there)

File: examples/customresources/demos/simplemodule/module.go
Comment (benjirewis) [reply examples/customresources/demos/simplemodule/module.go:60]: Updated those examples further to show what your suggesting formatting is now producing (no quote around the message and `\t` between caller and message).

File: robot/client/client.go
Comment (benjirewis) [reply robot/client/client.go:950]: Ah thanks! Done.

File: robot/client/client.go
Comment (benjirewis) [reply robot/client/client.go:954]: Cool; filed RSDK-6280 and referenced it here.

File: robot/server/server.go
Comment (benjirewis) [reply robot/server/server.go:421]: Yep! Noted in RSDK-6280 and put a `TODO` to preserve timezones above `moduleLogger`'s declaration.

File: robot/server/server.go
Comment (benjirewis) [reply robot/server/server.go]: Ah that makes sense, good catch! Added a similar line back in and explained its presence in a comment.

**Comment by benjirewis** [comment]
Oops just saw the test failures; looking into it. Feel free to hold off on review.

File: module/testmodule/main.go
Comment (dgottlieb) [inline module/testmodule/main.go:145]: I understand this isn't a real module, but these can now be `CDebug(ctx, ...)` or `CInfo(ctx, ...)`

File: examples/customresources/demos/simplemodule/module.go
Comment (benjirewis) [reply examples/customresources/demos/simplemodule/module.go:63]: > "counter" is a component this module can create.

Yep!

> Can we have newCounter or counter.Reconfigure log something to see that it gets the right namespace as well?

From some local testing, the logger namespace looks correct. Logs look like:

```
2024-01-17T11:56:00.016-0500	INFO	robot_server.simple-module.rdk:component:generic/counter1	simplemodule/module.go:63	logging from newCounter	{"module_log_ts": "2024-01-17T16:56:00.016Z"}
...
2024-01-17T11:59:31.301-0500	INFO	robot_server.simple-module.rdk:component:generic/counter1	simplemodule/module.go:79	logging from counter.Reconfigure	{"module_log_ts": "2024-01-17T16:59:31.295Z"}
```

File: module/testmodule/main.go
Comment (benjirewis) [reply module/testmodule/main.go:145]: Yep!

File: module/module.go
Comment (benjirewis) [inline module/module.go:262]: Made a decision here (see `NOTE`) due to failing tests that don't have any `parent` to connect to. Let me know what you think.

Comment (benjirewis) [reply module/module.go:262]: Trying to get rid of this logic (and simply return an error in the event that connection establishment back to the parent fails) by using a mock robot server in a few different tests; please hold off on reviewing.

File: logging/appender.go
Comment (maximpertsov) [inline logging/appender.go:13]: [q] is the meant to editable by users? if so I'd prefer to rename this to just `TimeFormat`. 

otherwise:

```suggestion
const DefaultTimeFormatStr = "2006-01-02T15:04:05.000Z0700"
```

File: module/log.go
Comment (maximpertsov) [inline module/log.go:72]: let's make this argument position-less via https://pkg.go.dev/flag or something similar. i feel like a user could easily mis-configure this and then be confused why the log level is not `debug` as they intended

this can be a follow-up since the behavior was already present

File: module/module.go
Comment (maximpertsov) [inline module/module.go:260]: [minor] no need to attribute since we have git blame - just a little less noise :)

```suggestion
	// moduleLoggers may be creating the client connection
```

**Review by maximpertsov — COMMENTED** [review]
left some review comments up until I reached https://github.com/viamrobotics/rdk/pull/3333/files#r1480621387

will hold off until it's ready again

File: module/module.go
Comment (benjirewis) [reply module/module.go:262]: Got rid of this logic; we now error in the event that connection establishment back to parent fails and I'm now using a mock robot service in a new `testutils/module_utils.go` file. Feel free to review now.

File: module/log.go
Comment (benjirewis) [reply module/log.go:72]: Ah that's a good idea; I think I'll push to another ticket/PR if that SGTY.

File: logging/appender.go
Comment (benjirewis) [reply logging/appender.go:13]: It is not meant to be editable; good catch! I'll remove the `Default` language.

File: module/example_test.go
Comment (benjirewis) [inline module/example_test.go:59]: I'm torn on what to do with this `example_test`. See how [this appears in our documentation](https://pkg.go.dev/go.viam.com/rdk@v0.20.0/module#example-package). In order to make `checkReady` work, I need to set up a fake robot (or a real one?) that listens at a parent address. I added this in, but felt like it made the example more complicated/confusing. I know that `checkReady` is trying to test the functionality of the module we created, but we have tests for this already. Curious what others think of me removing this part of the example.

File: module/module.go
Comment (benjirewis) [inline module/module.go:256]: Found this issue from `TestModuleFunctions` which sends two `ReadyRequest`s concurrently. If an `rdk` were to retry sending a `ReadyRequest` to a module for any reason, I think we could feasibly make a second connection back to the parent and leak the first, so I just disallowed the setting of `parent` more than once.

File: module/module_test.go
Comment (benjirewis) [inline module/module_test.go:191]: We can no longer send `ReadyRequest`s without a `ParentAddress` to modules in tests.

File: module/example_test.go
Comment (maximpertsov) [reply module/example_test.go:59]: yeah that's annoying... is this example more for testing or documentation? if the former then I'm okay to remove. if the latter, it depends on what functionality we are trying to demonstrate to the user with this `checkReady` bit - is that functionality documented by another more complex example?

File: logging/appender.go
Comment (benjirewis) [reply logging/appender.go]: My bad 😮‍💨 , missed that you were suggesting `const` and a potential rename. I'll keep `DefaultTimeFormatStr` and make it `const`.

File: module/example_test.go
Comment (dgottlieb) [reply module/example_test.go:59]: Benji and I talked and I'm also a bit torn on the best way to keep the functionality, but also make the example not look bloated.

I will mention what wasn't obvious to me -- this "check ready" is expected to print out the APIs and models the module registered. It does this by calling the gRPC engpoint for the `ReadyRequest` which returns this information. And, "check ready" is pretending to be a viam-server. Pre-patch, pretending was fine -- there was no requirement to make send a ready request. But now there is a requirement.

 We're in this (small) pickle because the `ReadyRequest` handles both the lower-level network initialization as well as the higher level viam-server<->module initialization. Different implementations of module startup could have avoided this by not bundling the two mechanisms into a single grpc call.

@cheukt, I suppose we don't already have a grpc endpoint for "return the API + model pairs" that's not the `ReadyRequest`?

File: logging/appender.go
Comment (maximpertsov) [reply logging/appender.go]: no worries! i said it in a confusing way. the main thing is making it a `const` if it's not meant to editable - either name is reasonable

File: module/module.go
Comment (dgottlieb) [reply module/module.go:256]: Can you add this information to the documentation? I'm fine keeping the nil-check for a test. Just want a reader to know `connectParent` will be called once per-ReadyRequest. And serving multiple `ReadyRequest`s is only expected to happen in tests.

File: logging/appender.go
Comment (benjirewis) [reply logging/appender.go]: > is there any way for users to set a different date format currently?

There is no way currently! Could file a ticket if that seems like a useful feature.

File: module/example_test.go
Comment (maximpertsov) [reply module/example_test.go:59]: I would vote to remove this example altogether for a few reasons:
* Documenting `checkReady` might give module author the idea that they need to do this validation, but they actually they don't since that is done by our module system AFAIK.
* The remaining functionality is already documented by our [simple module example](https://github.com/viamrobotics/rdk/blob/145fcda002054f6c04bd713c45a3cb92b1d546b3/examples/customresources/demos/simplemodule/module.go).
* As @benjirewis mentioned, we also [test the simple module](https://github.com/viamrobotics/rdk/blob/145fcda002054f6c04bd713c45a3cb92b1d546b3/module/modmanager/manager_test.go#L33-L138), so this example test is duplicative.

Comment (benjirewis) [reply module/example_test.go:59]: > I suppose we don't already have a grpc endpoint for "return the API + model pairs" that's not the ReadyRequest?

I don't think there is; here's the [module.proto](https://github.com/viamrobotics/api/blob/main/proto/viam/module/v1/module.proto) API.

I'm on the same side as @maximpertsov and would rather remove the example entirely, but we _can_ just remove the `checkReady` code if either of @dgottlieb or @cheukt want to leave the example test.

File: robot/client/client.go
Comment (benjirewis) [reply robot/client/client.go:950]: > Do you recall what the challenge was with that?

The challenge mostly had to do with conversion from `LogRequest` proto on the the server side (in `robot/server/server.go#Log`) to a "writable" log through `l.Info` (or the appropriate level). The code was non-trivial so I was hoping to kick it out to RSDK-6280. The eventual code would look something like the [`log_models` code](https://github.com/viamrobotics/app/blob/main/data/logs_models.go#L72-L143) in `app` with some extra code to actually "write" the log via `l.Info`. Not much comes "free" unfortunately, and it was  easier to put as much as possible in the `Message` field.

File: logging/appender.go
Comment (maximpertsov) [reply logging/appender.go]: cool, good to know! no need to make a ticket since it doesn't seem like anyone has been asking for it. I was just curious for my own understanding

File: module/module.go
Comment (benjirewis) [reply module/module.go:256]: Yep! Can add more of my original comment to the inline comment.

**Review by maximpertsov — APPROVED** [review]
aside from [this on-going discussion](https://github.com/viamrobotics/rdk/pull/3333#discussion_r1482161036), the rest of the changes LGTM. Happy to review again once we land on a decision.

File: module/example_test.go
Comment (dgottlieb) [reply module/example_test.go:59]: I'm good with removing altogether.

File: robot/client/client.go
Comment (dgottlieb) [reply robot/client/client.go:950]: Ah cool. Not just serializing the fields (and friends) -> proto. But also going from proto -> fields on the receiving end. Makes sense, tanks!

**Review by dgottlieb — APPROVED** [review]
Giving my approval regardless of where `example_test.go` lands.

File: robot/client/client.go
Comment (cheukt) [inline robot/client/client.go:947]: would change name to `Log` after api

Comment (benjirewis) [reply robot/client/client.go:947]: While this function does leverage the `Log` endpoint, it _is_  sort of specific to module logging. See the usage of the `moduleLogTSKey` and the leaving out of `Host` because it's not meaningful for modules. If you feel strongly that this should be named `Log`, I might make it more generic by adding a `forModule bool` param or something to attach the `moduleLogTSKey`. Thoughts?

File: module/log.go
Comment (benjirewis) [reply module/log.go:72]: Filed https://viam.atlassian.net/browse/RSDK-6624.

File: robot/client/client.go
Comment (cheukt) [reply robot/client/client.go:947]: I think making it more generic would be ideal, but not through using a `forModule` flag. Can we just add moduleLogTSKey as an extra field?

Comment (benjirewis) [reply robot/client/client.go:947]: 😮 that's a great idea. Done! Let me know if I interpreted that correctly.

File: module/module_test.go
Comment (benjirewis) [inline module/module_test.go:487]: `module/example_test.go` [defined](https://github.com/viamrobotics/rdk/blob/main/module/example_test.go#L21) a globally scoped `ctx` var set to `context.Background()` that a few other tests in the `module` package were relying on. I just added `ctx := context.Background()` to all of those tests.

**Review by cheukt — APPROVED** [review]
LGTM!

**Pull request merged**
